core/interfaces/i_draggable.ts

```diff
@@ -6,9 +6,47 @@
 
 // Former goog.module ID: Blockly.IDraggable
 
-import type {IDeletable} from './i_deletable.js';
+import {Coordinate} from '../utils/coordinate';
+import {IDragTarget} from './i_drag_target';
 
 /**
- * The interface for an object that can be dragged.
+ * Represents an object that can be dragged.
  */
-export interface IDraggable extends IDeletable {}
+export interface IDraggable {
+  /** Returns true iff the element is currently movable. */
+  isMovable(): boolean;
+
+  /** Returns the current location of the draggable in workspace coordinates. */
+  getLocation(): Coordinate;
+
+  /**
+   * Handles any drag startup (e.g moving elements to the front of the
+   * workspace).
+   *
+   * @param e PointerEvent that started the drag; could be used to
+   *     check modifier keys, etc.  May be missing when dragging is
+   *     triggered programatically rather than by user.
+   */
+  startDrag(e?: PointerEvent): void;
+
+  /**
+   * Handles moving elements to the new location, and updating any
+   * visuals based on that (e.g connection previews for blocks).
+   *
+   * @param newLoc Coordinate to which the draggable has been dragged.
+   * @param target The drag target the pointer is over, if any.
+   * @param e PointerEvent that started the drag; could be used to
+   *     check modifier keys, etc.  May be missing when dragging is
+   *     triggered programatically rather than by user.
+   */
+  drag(newLoc: Coordinate, target: IDragTarget | null, e?: PointerEvent): void;
```

Comment:
The style guide doesn't have guidance about whether to use `null` or `undefined` in general. But I would definitely prefer one or the other within an interface. Personally I would switch `target` to be optional.